colossalai/inference/tensor_parallel/engine.py

```diff
@@ -0,0 +1,254 @@
+        assert self.max_batch_size <= 64, "Max batch size exceeds the constraint"
+        assert self.max_input_len + self.max_output_len <= 2048, "Max length exceeds the constraint"
+
+        torch.device(device=device)
+        self.dtype = dtype
+
+        self.head_dim = self.model.config.hidden_size // self.model.config.num_attention_heads
+        self.head_num = self.model.config.num_attention_heads
+        self.layer_num = self.model.config.num_hidden_layers
+
+        self.tp_size = -1    # to be set with given shard config in self.prepare_shard_config
+        self.cache_manager = None
+
+    def _init_manager(self) -> None:
+        assert self.tp_size >= 1, "TP size not initialized without providing a valid ShardConfig"
+        assert self.head_num % self.tp_size == 0, f"Cannot shard {self.head_num} heads with tp size {self.tp_size}"
+        self.head_num //= self.tp_size    # update sharded number of heads
+        self.cache_manager = MemoryManager(self.max_total_token_num, self.dtype, self.head_num, self.head_dim,
+                                           self.layer_num)
+
+    def prepare_with_shard_config(self, shard_config: Optional[ShardConfig] = None) -> ShardConfig:
+        """ Prepare the engine with a given ShardConfig, or create a default one with tp size 1 """
+        self.tp_size = 1
+        if shard_config is None:
+            shard_config = ShardConfig(
+                tensor_parallel_process_group=None,
+                pipeline_stage_manager=None,
+                enable_tensor_parallelism=False,
+                enable_fused_normalization=False,
+                enable_all_optimization=False,
+                enable_flash_attention=False,
+                enable_jit_fused=False,
+                inference_only=True,
+            )
+        else:
+            shard_config.inference_only = True
+            shard_config.pipeline_stage_manager = None
+            if shard_config.enable_tensor_parallelism:
+                self.tp_size = shard_config.tensor_parallel_size
+        self._init_manager()
+
+        return shard_config
+
+    def shard_model_by(self, shardformer: ShardFormer) -> None:
+        """ Shard the model and store the sharded model by given ShardFormer """
+        assert self.tp_size == shardformer.shard_config.tensor_parallel_size, \
+            "Discrepancy between the tp size of TPInferEngine and the tp size of shard config"
+        model_name = self.model.__class__.__name__
+        assert model_name in self._supported_models(), f"Unsupported model cls {model_name} for TP inference."
+        policy = get_autopolicy(self.model, inference_only=True)
+        self.sharded_model, _ = shardformer.optimize(self.model, policy)
+        self.sharded_model = self.sharded_model.cuda()
+
+    @staticmethod
+    def _supported_models() -> List[str]:
+        return _supported_models
+
+    def generate(self, input_tokens, generate_kwargs) -> torch.Tensor:
```

Comment:
What's the type of `input_tokens`?

Reply:
`**generate_kwargs` may be better

Reply:
input_tokens could be BatchEncoding, torch tensor, or list. Type hints added and kwargs revised.

Reply:
done